Performant pagination during bulk post indexing

WPprodigy opened the pull request:
# Purpose

Pagination with SQL's offset can get considerably slow on large sites as it has to tread through all the rows until it finds where to pick up. This PR switches away from using offset and instead ranges of post IDs which allows it to execute queries much faster. I've also prevented the need for running `SQL_CALC_FOUND_ROWS` on each query. There are comparisons of the query changes and performance profiling at the end.

To accomplish this new way of pagination while still using WP_Query, I'm passing custom query args and then filtering on `posts_where` based on these args.

# Testing

At the moment, I've kept the functionality behind a feature flag to make testing/comparing easier. To test, you'll want to apply this PR but then also add the `[--advanced-pagination] [--start-post-id]` CLI args to the docblock in `search/includes/classes/commands/class-corecommand.php` to allow them to passthrough the `wp vip-search index` CLI if you want to use that.

Ideally this just happens by default when it's possible, and we remove the feature flag & related logic when it's merged.

```
wp vip-search index --version=2 --indexables=post --advanced-pagination --start-post-id=2500000
```

I'm not 100% sure on the "end range" functionality you'll see in the code, and would appreciate some advice there. Mainly, is it worth it? Removing the end range appears to actually (very slightly) speed up the query and makes the code a bit simpler, but having the end-range does highly limit the rows scanned. Could go even further on limiting rows scanned, as right now there is an arbitrary limit of 50,000 IDs. In testing, changing this to lower doesn't really have an effect - it's already very fast.

# Profiling

```
### CURRENT STATUS

explain SELECT SQL_CALC_FOUND_ROWS  wp_posts.* FROM wp_posts  WHERE 1=1  AND wp_posts.post_type IN (XXX) AND ((wp_posts.post_status = 'publish' OR wp_posts.post_status = 'future' OR wp_posts.post_status = 'trash' OR wp_posts.post_status = 'inherit' OR wp_posts.post_status = 'request-pending' OR wp_posts.post_status = 'request-confirmed' OR wp_posts.post_status = 'request-failed' OR wp_posts.post_status = 'request-completed' OR wp_posts.post_status = 'spam' OR wp_posts.post_status = 'pitch' OR wp_posts.post_status = 'draft' OR wp_posts.post_status = 'needs-editing' OR wp_posts.post_status = 'ready-to-publish' OR wp_posts.post_status = 'sync-only' OR wp_posts.post_status = 'private'))  ORDER BY wp_posts.ID DESC LIMIT 1500000, 500;

+------+-------------+----------+-------+------------------+---------+---------+------+---------+-------------+
| id   | select_type | table    | type  | possible_keys    | key     | key_len | ref  | rows    | Extra       |
+------+-------------+----------+-------+------------------+---------+---------+------+---------+-------------+
|    1 | SIMPLE      | wp_posts | index | type_status_date | PRIMARY | 8       | NULL | 5714850 | Using where |
+------+-------------+----------+-------+------------------+---------+---------+------+---------+-------------+

In the following tests, I had to drop to 250 posts per page instead of 500 (timeouts otherwise) and I removed SQL_CALC_FOUND_ROWS for a more fair comparison due to how I changed the logic in this PR.
Otherwise, you could expect 60+ second queries that often timeout. You'll see how offset gets progressively worse even with these adjustments:

Offset 2500: 0.35 secs
Offset 25000: 1.54 secs
Offset 250000: 23.42 secs
Offset 2500000: 35.66 sec

### WITH THIS PR

explain SELECT wp_posts.* FROM wp_posts  WHERE 1=1 AND wp_posts.ID < 2598793 AND wp_posts.ID > 2548793  AND wp_posts.post_type IN (XXX) AND ((wp_posts.post_status = 'publish' OR wp_posts.post_status = 'future' OR wp_posts.post_status = 'trash' OR wp_posts.post_status = 'inherit' OR wp_posts.post_status = 'request-pending' OR wp_posts.post_status = 'request-confirmed' OR wp_posts.post_status = 'request-failed' OR wp_posts.post_status = 'request-completed' OR wp_posts.post_status = 'spam' OR wp_posts.post_status = 'pitch' OR wp_posts.post_status = 'draft' OR wp_posts.post_status = 'needs-editing' OR wp_posts.post_status = 'ready-to-publish' OR wp_posts.post_status = 'sync-only' OR wp_posts.post_status = 'private'))  ORDER BY wp_posts.ID DESC LIMIT 0, 500;

+------+-------------+----------+-------+--------------------------+---------+---------+------+-------+-------------+
| id   | select_type | table    | type  | possible_keys            | key     | key_len | ref  | rows  | Extra       |
+------+-------------+----------+-------+--------------------------+---------+---------+------+-------+-------------+
|    1 | SIMPLE      | wp_posts | range | PRIMARY,type_status_date | PRIMARY | 8       | NULL | 81838 | Using where |
+------+-------------+----------+-------+--------------------------+---------+---------+------+-------+-------------+

 Actual Query Time: 0.28 sec. No noticeable fluctuation with differing IDs.

Perhaps worth noting that query time is 0.27 sec if I remove the end range (wp_posts.ID > 2548793), though then 2857433 rows are scanned.
```

rinatkhaziev reviewed (changes requested):
This is exciting and should help a lot! Thanks for working on it! Looks good, I left some feedback inline.

As for the end range, I'd generally favor simplicity but not at the cost of less stability. My limited testing on a site with several million posts confirmed your findings in regards to the performance of having end range vs not having it (about the same).

Additionally, I think we need a hard cut-off argument (`--end-post-id`). E.g. I want to index all posts with IDs between 1500 and 1000 and I don't care about the rest. Not having that argument would force me to either run the indexing all the way or babysitting the command to terminate it (which may not be feasible if a command runs in a context that doesn't allow a user to interrupt it).

WPprodigy commented on includes/classes/Command.php:
```diff
@@ -747,11 +746,21 @@ function ( $prefix ) use ( $args ) {
+			if ( ! empty( $args['start-post-id'] ) ) {
+				$query_args['ep_indexing_start_post_id'] = absint( $args['start-post-id'] );
+				// Perpetually remember for the "total_objects" count to keep the progress bar accurate.
+				$query_args['ep_indexing_initial_start_post_id'] = absint( $args['start-post-id'] );
```
Not sure I follow. So every round there is a progress message like: `Processed 500/300,000`. We get that 300,000 from the first query that runs with `SQL_CALC_FOUND_ROWS` present.

Say there are 250,000 posts that need to be indexed, and the CLI is passed with `--start-post-id=100000`. The first loop will say `Processed 500/150,000`, the next loop - `Processed 1000/150,000`. 

It needs to remember that initial `start-post-id` that is passed in so that the first `SQL_CALC_FOUND_ROWS` query that tells us how many posts are being indexed during this run is always the same (and thus cached).

rinatkhaziev replied:
Ohh I see, I read this comment as "use this value as total_objects count" and got confused :)

WPprodigy replied:
Switching the logic to be based around a moving `ep_indexing_last_processed_object_id` value instead. So `start_post_id` will always stay the same based on initial input - thus no more need for `ep_indexing_initial_start_post_id`.

WPprodigy commented on includes/classes/Indexable/Post/Post.php:
```diff
@@ -95,14 +97,65 @@ public function query_db( $args ) {
 		$query = new WP_Query( $args );
 
+		// If the query failed to return any posts, remove the second part of the range to ensure there truly are no matching posts left.
+		if ( empty( $query->posts ) && ! empty( $args['ep_indexing_advanced_pagination'] ) ) {
```
So let's say the site has post IDs 1-1,000 of normal posts. Then nothing-ness until ID 250,000. The query with end ranges will be like `WHERE wp_posts.ID > 1000 AND wp_posts.ID < 51000`. This will return no posts, causing the abstracted logic in `index_helper()` to abort out of the loop because it thinks we're out of posts to index.

WPprodigy replied:
Oh wait, I think I misunderstood and this comment was about line 129 instead. In that case though, I'm not sure of the best option for this still due to the abstraction layer.

WPprodigy replied:
Added a second function param of `$exclude_total_objects_count` to dictate whether or not we need to fetch the total object count. Then in Command.php we're now only asking for the total object count once, before the loop begins.

Notably, this is backwards compatible with all the other places that use query_db() and still rely on getting that total objects count.

WPprodigy commented on includes/classes/Indexable/Post/Post.php:
```diff
@@ -95,14 +97,65 @@ public function query_db( $args ) {
+		$start_post_id = $query->get( 'ep_indexing_start_post_id', null );
+
+		if ( $using_advanced_pagination && is_int( $start_post_id ) ) {
+			$cutoff_point_id = max( $start_post_id - 50000, 0 );
```
That's a good point about max_join_size, hadn't though of thought. It is odd it doesn't come up in the $wpdb tests, but alas seems fine to continue with the end range. Especially if adding support for passing `--end-post-id` anyways 👍

WPprodigy replied:
Actually, abandoning the artificial end ranges for now. More details in latest comment.

WPprodigy commented on includes/classes/Command.php:
```diff
@@ -897,8 +908,11 @@ function ( $prefix ) use ( $args ) {
 
-			if ( ! $no_bulk ) {
-				WP_CLI::log( sprintf( esc_html__( 'Processed %1$d/%2$d...', 'elasticpress' ), (int) ( count( $query['objects'] ) + $query_args['offset'] ), (int) $query['total_objects'] ) );
+			WP_CLI::log( sprintf( esc_html__( 'Processed %1$d/%2$d. Last Post ID: %3$d', 'elasticpress' ), (int) ( $synced + count( $failed_objects ) ), (int) $query['total_objects'], (int) $query_args['ep_indexing_start_post_id'] ) );
```
Hah, I noticed this earlier today as well! I'll rename the text here and also switch the variables over to a more generic `ep_indexing_start_object_id`.

WPprodigy commented on includes/classes/Indexable/Post/Post.php:
```diff
@@ -45,6 +45,8 @@ public function __construct() {
 		$this->sync_manager      = new SyncManager( $this->slug );
 		$this->query_integration = new QueryIntegration( $this->slug );
+
+		add_filter( 'posts_where', array( $this, 'bulk_indexing_filter_posts_where' ), 9999, 2 );
```
I agree it's not super great, but also can't find a better option :/. 

- `Command->index_helper()` is technically when the bulk index query is kicked off and where it finishes, though this is object agnostic and it would be even more "off" doing this there.
- Could check if the filter needs to be added during `Indexable\Post->query_db()` perhaps and do so before the query runs, though that's logic that will run a lot more than where the filter currently sites.

What are you thinking would be better than how it currently is?

WPprodigy replied:
Gonna go with option 2. This is indeed a problem during tests, was ending up with:

```
AND wptests_posts.ID <= 9 AND wptests_posts.ID >= 0 AND wptests_posts.ID <= 9 AND wptests_posts.ID >= 0
```

WPprodigy commented on includes/classes/Indexable/Post/Post.php:
```diff
@@ -95,14 +97,65 @@ public function query_db( $args ) {
+		// Implement custom pagination for better performance.
+		if ( ! empty( $args['ep_indexing_advanced_pagination'] ) ) {
+			// Apply an initial start_post_id if needed.
+			$args['ep_indexing_start_post_id'] = ! empty( $args['ep_indexing_start_post_id'] ) ? $args['ep_indexing_start_post_id'] : PHP_INT_MAX;
```
Discussed in slack, but just noting here for historical reasons that WP is pretty strict about posts not going over PHP_INT_MAX. https://github.com/WordPress/WordPress/blob/d0a32c51112bc4f4827476ad417cc8905abcd85c/wp-includes/class-wp-post.php#L234-L237

Going to stick with `PHP_INT_MAX` as the default for now, though I did switch most of the logic to be more supportive in case this is changed in the future.

WPprodigy commented on includes/classes/Command.php:
```diff
@@ -897,8 +908,11 @@ function ( $prefix ) use ( $args ) {
+
+			$loop_counter++;
+			if ( ( $loop_counter % 10 ) === 0 ) {
+				WP_CLI::log( WP_CLI::colorize( '%Y' . esc_html__( 'Time elapsed: ', 'elasticpress' ) . '%N' . timer_stop() ) );
```
Added. Logging looks like this now:

```
Time elapsed: 11.26 (+5.75)
Memory Usage: 60.34mb (Peak: 60.46mb)
```

WPprodigy commented:
Addressed the feedback. Some notable changes:

- Cleaned up and made this behavior the default, along with proper support for passing in `--start-object-id` along with `--end-object-id`. (Will now actually index the start id rather than the one after it)
- Reworked the query arg logic so that `last_processed_object_id` is now what we pass between loops. This will allow for greater flexibility in the future, such as if we wanted to switch up the order to ASC for example. All of the related logic for the ranges now lives in `bulk_indexing_filter_posts_where()` and there alone (quite abstract, could be used for terms later on).
- Logging enhancements.

I abandoned the artificial end ranges for now, due to complexities with how the index_helper() loop is structured / how it decided when to stop looping. Still not completely sure why queries through $wpdb don't trigger the MAX_JOIN_SIZE error, but seems to be alright and query time is actually the slightest bit faster without the end range for whatever reason.

WPprodigy commented on includes/classes/Indexable/Post/Post.php:
```diff
@@ -63,46 +64,111 @@ public function query_db( $args ) {
 		 * @return  {array} New arguments
 		 */
-		$args = apply_filters( 'ep_index_posts_args', $args );
+		$args = apply_filters( 'ep_post_query_db_args', apply_filters( 'ep_index_posts_args', wp_parse_args( $args, $defaults ) ) );
```
Will flip these two filters around to respect the previous order of events.

WPprodigy commented:
Alright, think it's in a good place now :)

rinatkhaziev commented on includes/classes/Indexable/Post/Post.php:
```diff
@@ -73,36 +75,114 @@ public function query_db( $args ) {
+			// On the first loopthrough we begin with the requested start ID. Afterwards, use the last processed ID to paginate.
+			$start_range_post_id = $requested_start_id;
+			if ( is_numeric( $last_processed_id ) ) {
+				$start_range_post_id = number_format( $last_processed_id - '1', 0, '.', '' );
```
Since we fell back to using strictly integers because WP can't handle IDs > PHP_INT_MAX properly let's simplify this to `$last_processed_id - 1` as there's no need in number_format workaround anymore.
```suggestion
				$start_range_post_id =  $last_processed_id - 1;
```

rinatkhaziev reviewed (changes requested):
Thanks for the changes and refactors! I left a minor note/suggestion re number_format we discussed.

rinatkhaziev reviewed (approved):
:shipit:
